miniapps/tools/least-squares.cpp

```diff
@@ -0,0 +1,1720 @@
+   MFEM_VERIFY(params.ord_smooth > params.ord_src,
+               "Smooth space must be more regular!");
+   MFEM_VERIFY((0 <= params.solver_type) && (params.solver_type < num_solvers),
+               "Invalid solver type: " << params.solver_type);
+   MFEM_VERIFY((0 <= params.reg_type) && (params.reg_type < num_regularization),
+               "Invalid regularization type: " << params.reg_type);
+
+   if (Mpi::Root())
+   {
+      mfem::out << "Number of serial refs.:  " << params.ser_ref << "\n"
+                << "Number of parallel refs: " << params.par_ref << "\n\n"
+                << "Order smooth FES:        " << params.ord_smooth << "\n"
+                << "Order source FES:        " << params.ord_src << "\n"
+                << "Order destination FES:   " << params.ord_dst << "\n\n"
+                << "Newton relative tol:     " << params.newton.rtol << "\n"
+                // << "Newton absolute tol:     " << params.newton.atol << "\n"
+                << "Newton max. num. iter.:  " << params.newton.max_iter << "\n\n"
+                << std::endl;
+   }
+
+   // Mesh
+   const int num_x = 2;
+   const int num_y = 2;
+   Mesh serial_mesh = Mesh::MakeCartesian2D(num_x, num_y, Element::QUADRILATERAL);
+   for (int i = 0; i < params.ser_ref; ++i) { serial_mesh.UniformRefinement(); }
+   serial_mesh.EnsureNCMesh();
+   ParMesh mesh(MPI_COMM_WORLD, serial_mesh);
+   for (int i = 0; i < params.par_ref; ++i) { mesh.UniformRefinement(); }
+   mesh.EnsureNCMesh();
+
+   // target function u(x,y)
+   std::function<real_t(const Vector &)> u_function;
+   switch (params.example)
+   {
+      default:
+      case plane:
+         u_function = example::plane;
+         break;
+      case sine:
+         u_function = example::sine;
+         break;
+      case exp_sine:
+         u_function = example::exp_sine;
+         break;
+   }
+   FunctionCoefficient u_coefficient(u_function);
+
+   // Broken spaces
+   L2_FECollection fec_smooth(params.ord_smooth, mesh.Dimension());
+   L2_FECollection fec_src(params.ord_src, mesh.Dimension());
+   L2_FECollection fec_dst(params.ord_dst, mesh.Dimension());
+
+   ParFiniteElementSpace fes_smooth(&mesh, &fec_smooth);
+   ParFiniteElementSpace fes_src(&mesh, &fec_src);
+   ParFiniteElementSpace fes_dst(&mesh, &fec_dst);
+
+   // Grid Functions
+   ParGridFunction u_smooth(&fes_smooth);
+   ParGridFunction u_src(&fes_src);
+   ParGridFunction u_dst_avg(&fes_src);
+   ParGridFunction diff(&fes_src);
+   ParGridFunction u_dst(&fes_dst);
+
+   u_smooth.ProjectCoefficient(u_coefficient);
+   // TODO(Gabriel): What if fes_src is H.O.?
+   u_smooth.GetElementAverages(u_src);
+
+   // Solver choice
+   // TODO(Gabriel): Support more PCs?
+   /* Notes:
+    * OperatorJacobiSmoother will call AssembleDiagonal on
+    * Operator, even if the Setup function will be called
+    * later on. DenseMatrix does not have this method
+    * implemented.
+    */
+   std::shared_ptr<Solver> solver(nullptr);
```

Comment:
Technically, I think the trivial constructor already sets the object to `nullptr`, so you can replace with `std::shared_ptr<Solver> solver`... although I would replace `solver` with the more descriptive `linear_solver`